synapse/storage/engines/sqlite.py

```diff
@@ -135,13 +135,14 @@ def executescript(cursor: sqlite3.Cursor, script: str) -> None:
         > than one statement with it, it will raise a Warning. Use executescript() if
         > you want to execute multiple SQL statements with one call.
 
-        Though the docs for `executescript` warn:
+        The script is wrapped in transaction control statemnets, since the docs for
+        `executescript` warn:
 
         > If there is a pending transaction, an implicit COMMIT statement is executed
         > first. No other implicit transaction control is performed; any transaction
         > control must be added to sql_script.
         """
-        cursor.executescript(script)
```

Comment:
Unfortunately, `script` will not contain the statement https://github.com/matrix-org/synapse/blob/0a38c7ec6d46b6e51bfa53ff44e51637d3c63f5c/synapse/storage/prepare_database.py#L540-L544
which adds a row to `applied_schema_deltas`.

It's possible that we are interrupted after running the migration and the COMMIT added here, but before we mark the transaction as done. Then on the next restart, we'll try to reapply an applied migration, which I expect will fail. It's a small window, and having the COMMIT here is a vast improvement... so I think we should probably leave it as-is?

Reply:
That's a good point.

I wasn't sure how a `BEGIN` without a corresponding `COMMIT` would behave in `executescript`. The implicit commit at the start implies that some level of magic is going on.

Reply:
Could we insert that as the last line of the `script`?

Reply:
We could, it'll just take a bunch of refactoring. I'm tempted to omit the final COMMIT and add some unit tests instead.

Reply:
executescript is defined here:
https://github.com/python/cpython/blob/7e3f09cad9b783d8968aa79ff6a8ee57beb8b83e/Modules/_sqlite/cursor.c#L101

and it looks like there's no reason leaving the transaction open shouldn't work.